docs: add kor/eng metallb.md

kosehy opened the pull request:
## Changes?
<!-- 이 pr 로 인해서 무엇이 변경되었는지 작성해주세요 -->
- contributors 에 seunghyun-ko 폴더를 추가했습니다.
- appendix에 베어메탈 환경에서 사용할 수 있는 로드 벨런서 MetalLB에 대한 한글/영어 문서를 추가했습니다.

## Why we need?
<!-- 이 pr 이 왜 필요한지 작성해주세요 -->
 MetalLB를 사용하면 kubectl port-forward 를 사용하지 않고 사용자가 원하는 IP 주소로 쿠버네티스 서비스에 접근할 수 있습니다. 
## Test?

- [x] `npm run start` 를 수행하여 로컬에서 렌더링된 페이지를 확인하셨나요?
- [x] `npm test` 를 통과하였나요?

## Anything Else? (Optional)
<!-- 스크린샷, 환경 정보, 주의사항 등 필요한 추가정보가 있다면 작성해주세요. -->

kosehy committed:
docs: add metallb loadbalancer documentation

Update content/kor/docs/appendix/metallb.md

띄어쓰기, 문장 다듬기,  MetalLB 링크 추가 피드백 반영

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

엄격한 ARP 뒤에 용어 추가 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

demonset 오타 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

문장 다듬기 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

두 문장을 한 문장으로 합치는 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

문장 다듬기 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

문장 다듬기 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

문장 다듬기 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

Update content/kor/docs/appendix/metallb.md

editor 삭제 피드백 적용

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>

docs: apply feedbacks to make better documentation.

docs: apply feedbacks to en docs

docs: add two more metallb usecase

Add mlflow & Grafana usecase

Apply feedbacks from @Aiden-Jeon and @anencore94

Co-authored-by: Jaeyeon Kim <anencore94@gmail.com>
Co-authored-by: Jongseob Jeon <anencore94@gmail.com>

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+## MetalLB란?
+
+Kubernetes사용시 AWS, GCP, Azure 와 같은 클라우드 플랫폼에서는 로드 벨런서를 제공해 주지만 베어메탈 클러스터에 대한 네트워크 로드 밸런서 구현을 제공하지 않습니다.
+MetalLB는 베어메탈 환경에서 사용할 수 있는 로드 벨런서를 제공하는 오픈소스 프로젝트 입니다.
```
```suggestion
Kubernetes 사용 시 AWS, GCP, Azure 와 같은 클라우드 플랫폼에서는 자체적으로 로드 벨런서(Load Balancer)를 제공해 주지만, 온프레미스 클러스터에서는 로드 벨런싱 기능을 제공하는 모듈을 추가적으로 설치해야 합니다.
[MetalLB](https://metallb.universe.tf/)는 베어메탈 환경에서 사용할 수 있는 로드 벨런서를 제공하는 오픈소스 프로젝트 입니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+### Preparation
+
+IPVS 모드에서 kube-proxy를 사용하는 경우 Kubernetes v1.14.2 이후부터는 엄격한 ARP 모드를 사용하도록 설정해야 합니다.
```
```suggestion
IPVS 모드에서 kube-proxy를 사용하는 경우 Kubernetes v1.14.2 이후부터는 엄격한 ARP(strictARP) 모드를 사용하도록 설정해야 합니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
````diff
@@ -0,0 +1,280 @@
+sed -e "s/strictARP: false/strictARP: true/" | \
+kubectl apply -f - -n kube-system
+
+```
````
조금 더 쉬운 적용을 위해서 문장을 다음과 같이 변경하면 어떨지 제안드립니다.

---

엄격한 ARP 모드를 적용하기에 앞서, 현재 모드를 확인합니다.

```text
kubectl get configmap kube-proxy -n kube-system -o yaml | \
grep strictARP
```

`strictARP: false` 가 출력되는 경우 다음을 실행하여 `strictARP: true`로 변경합니다. (`strictARP: true`가 이미 출력된다면 다음 커맨드를 수행하지 않으셔도 됩니다.)

```text
kubectl get configmap kube-proxy -n kube-system -o yaml | \
sed -e "s/strictARP: false/strictARP: true/" | \
kubectl apply -f - -n kube-system
```

kosehy replied:
@anencore94 님의 주신 피드백 처럼 문장을 수정하겠습니다

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+metallb-system/controller 는 deployment 로 동작하며 IP 주소 할당을 처리하는 클러스터 전체 컨트롤러 입니다.
+
+metallb-system/speaker 는 demonset 형태로 배포되며, 외부 트래픽과 서비스를 연결해 네트워크 통신이 가능하도록 구성하는 역할을 담당합니다.
```
```suggestion
metallb-system/speaker 는 daemonset 형태로 배포되며, 외부 트래픽과 서비스를 연결해 네트워크 통신이 가능하도록 구성하는 역할을 담당합니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+매니페스트의 구성 요소는 다음과 같습니다.
+
+metallb-system/controller 는 deployment 로 동작하며 IP 주소 할당을 처리하는 클러스터 전체 컨트롤러 입니다.
```
```suggestion
metallb-system/controller 는 deployment 로 배포되며, 로드 벨런싱을 수행할 external IP 주소의 할당을 처리하는 역할을 담당합니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+Layer 2 모드는 정말 간단 합니다.
+
+사용할 IP 주소만 설정하면 됩니다.
```
```suggestion
Layer 2 모드의 사용법은 간단합니다. 사용할 IP 주소의 대역만 설정하면 됩니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+## Configuration
+
+MetalLB 설정은 ConfigMaps에 배포할 config 를 수정하여 설정 할 수 있습니다.
```
```suggestion
MetalLB 의 자세한 로드 벨런싱 정책 설정은 관련 설정 정보를 담아 k8s configmap 의 형태로 배포하여 설정할 수 있습니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
````diff
@@ -0,0 +1,280 @@
+      protocol: layer2
+      addresses:
+      - 192.168.35.100-192.168.35.110
+```
````
이 파일을 kubectl apply -f 로 배포하라는 안내가 추가되면 좋을 것 같습니다.

kosehy replied:
@anencore94 님의 주신 피드백 처럼 kubectl apply -f 로 배포하라는 안내를 추가하겠습니다.

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+### Kubeflow GUI
+
+먼저 kubeflow gui의 External-IP type 을 LoadBalancer 로 바꾸기전에 svc 를 확인합니다.
```
```suggestion
먼저 kubeflow의 UI를 제공하는 istio-system 네임스페이스의 istio-ingressgateway 서비스의 타입을 `LoadBalancer`로 변경하여 MetalLB로부터 로드 벨런싱 기능을 제공받기 전에, 현재 상태를 확인합니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
````diff
@@ -0,0 +1,280 @@
+kubectl get svc/istio-ingressgateway -n istio-system
+```
+
+External-IP 값이 none 인 것을 확인합니다.
````
```suggestion
해당 서비스의 타입은 ClusterIP이며, External-IP 값은 none 인 것을 확인할 수 있습니다.
```

anencore94 commented on content/kor/docs/appendix/metallb.md:
````diff
@@ -0,0 +1,280 @@
+istio-ingressgateway   ClusterIP   10.103.72.5   <none>        15021/TCP,80/TCP,443/TCP,31400/TCP,15443/TCP   4h21m
+
+```text
+type 을 LoadBalancer 로 변경하고
````
````text 를 새로 여시는 게 아니라 이전 ```text 를 닫으려하신게 맞나요?

kosehy replied:
맞습니다. text 단어를 삭제하겠습니다

anencore94 commented on content/kor/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+로컬 네트워크의 ARP 요청에 직접 응답하여 컴퓨터의 MAC주소를 클라이언트에 제공하는 방식으로 작동합니다.
+
+다음 예제에서는 MetalLB 가 192.168.35.100 ~ 192.168.35.110의 IP에 대한 제어 권한을 제공하고 Layer 2 모드를 구성합니다.
```
클러스터 노드와 클라이언트 노드가 분리된 경우를 위해, `192.168.35.100 ~ 192.168.35.110` 대역이 클라이언트 노드와 클러스터 노드 모두 접근 가능한 대역이어야 한다는 걸 추가해주어야 할 것 같습니다.

kosehy replied:
해당 문구를 추가하겠습니다.

anencore94 commented on content/kor/docs/appendix/metallb.md:
````diff
@@ -0,0 +1,280 @@
+추가 안 할 경우에는 위에서 설정했던 IP 주소풀에서 순차적으로 IP 주소가 배정됩니다.
+
+```text
+KUBE_EDITOR=nano kubectl edit svc/istio-ingressgateway -n istio-system
````
```suggestion
kubectl edit svc/istio-ingressgateway -n istio-system
```

editor는 강제하지 않는 것이 좋을 것 같습니다

anencore94 reviewed (commented):
좋은 의견 감사합니다! 여러 사용자를 위해 appendix 로 추가하신 부분도 센스가 넘치시네요 :+1: 
몇몇 코멘트를 남겨두었으니 확인해주시면 감사하겠습니다!

anencore94 commented:
`npm test` 확인해주시면 감사하겠습니다!

또한 package-lock.json 은 삭제하신 이유가 있으실까요? 제가 frontend 에 대해 잘 이해하고 있지는 않지만, package-lock.json 은 정확한 node package 버전 관리를 위해서는 git diff 가 많이 발생하더라도 소스코드와 함께 관리하는 방향이 best practice 인 것으로 주워들었습니다..ㅎㅎ
그래서 현재 github action 에서도 packge.json 과 함께 package-lock.json 을 바탕으로 패키지 버전을 고정시켜 페이지 렌더링이 되는 것을 확인하고 있습니다.(https://github.com/mlops-for-all/mlops-for-all.github.io/runs/4717966022?check_suite_focus=true)
혹시 제가 잘못 이해하고 있다면 말씀해주세요!

- https://dev.to/adamklein/package-lock-json-in-git-or-not-50l5
- https://hyunjun19.github.io/2018/03/23/package-lock-why-need/

kosehy commented:
> `npm test` 확인해주시면 감사하겠습니다!
> 
> 또한 package-lock.json 은 삭제하신 이유가 있으실까요? 제가 frontend 에 대해 잘 이해하고 있지는 않지만, package-lock.json 은 정확한 node package 버전 관리를 위해서는 git diff 가 많이 발생하더라도 소스코드와 함께 관리하는 방향이 best practice 인 것으로 주워들었습니다..ㅎㅎ 그래서 현재 github action 에서도 packge.json 과 함께 package-lock.json 을 바탕으로 패키지 버전을 고정시켜 페이지 렌더링이 되는 것을 확인하고 있습니다.(https://github.com/mlops-for-all/mlops-for-all.github.io/runs/4717966022?check_suite_focus=true) 혹시 제가 잘못 이해하고 있다면 말씀해주세요!
> 
> * https://dev.to/adamklein/package-lock-json-in-git-or-not-50l5
> * https://hyunjun19.github.io/2018/03/23/package-lock-why-need/

이 부분은 제가 착각해서 package-lock.json 파일을 삭제했네요...
해당 파일은 다시 추가하겠습니다.

> 좋은 의견 감사합니다! 여러 사용자를 위해 appendix 로 추가하신 부분도 센스가 넘치시네요 👍 몇몇 코멘트를 남겨두었으니 확인해주시면 감사하겠습니다!

@anencore94 자세히 리뷰해 주셔서 정말 감사합니다! @anencore94 님의 의견을 반영해서 다시 커밋하겠습니다.

Aiden-Jeon commented:
@kosehy 
안녕하세요!
이미지 경로가 틀려서 노출되지 않는 문제가 있습니다.
현재 이미지가 아래 경로에 있는데
`images/docs/metallb/login-after-istio-ingressgateway-setting.png`
이 쪽으로 옮겨주시면 될 것 같습니다!
`static/images/docs/metallb/login-after-istio-ingressgateway-setting.png`

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+- BGP 모드를 사용할 경우 BGP 기능을 지원하는 하나 이상의 라우터
+
+- memberlist에서 요구하는 대로 노드 간에 포트 TCP/UDP 7946 오픈
```
```suggestion
| 요구 사항                                                    | 버전 및 내용                                                 |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| Kubernetes                                                   | 로드 벨런싱 기능이 없는 >= v1.13.0                           |
| [호환가능한 네트워크  CNI](https://metallb.universe.tf/installation/network-addons/) | Calico, Canal, Cilium, Flannel, Kube-ovn, Kube-router, Weave  Net |
| IPv4 주소                                                    | MetalLB 배포에 사용                                          |
| BGP 모드를 사용할 경우                                       | BGP 기능을 지원하는 하나 이상의 라우터                       |
| 노드 간 포트 TCP/UDP 7946 오픈                               | memberlist 요구 사항                                         |
```
요구사항을 좀 더 확인하기 쉽도록 표로 만들었습니다.

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+# actually apply the changes, returns nonzero returncode on errors only
+kubectl get configmap kube-proxy -n kube-system -o yaml | \
+sed -e "s/strictARP: false/strictARP: true/" | \
+kubectl apply -f - -n kube-system
```
해당 부분에서 어떤 출력이 나와야 하는지 적어주시면 좋을 것 같습니다!

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
````diff
@@ -0,0 +1,280 @@
+```text
+kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.11.0/manifests/namespace.yaml
+kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.11.0/manifests/metallb.yaml
+
````
```suggestion
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.11.0/manifests/namespace.yaml
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.11.0/manifests/metallb.yaml
```

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+NAME                          READY   STATUS    RESTARTS   AGE
+controller-7dcc8764f4-8n92q   1/1     Running   1          1m
+speaker-fnf8l                 1/1     Running   1          1m
+
```
```suggestion
NAME                          READY   STATUS    RESTARTS   AGE
controller-7dcc8764f4-8n92q   1/1     Running   1          1m
speaker-fnf8l                 1/1     Running   1          1m
```

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+metallb-system/controller 는 deployment 로 동작하며 IP 주소 할당을 처리하는 클러스터 전체 컨트롤러 입니다.
+
+metallb-system/speaker 는 demonset 형태로 배포되며, 외부 트래픽과 서비스를 연결해 네트워크 통신이 가능하도록 구성하는 역할을 담당합니다.
```
```suggestion
- metallb-system/controller
  - deployment 로 배포되며, 로드 벨런싱을 수행할 external IP 주소의 할당을 처리하는 역할을 담당합니다.
- metallb-system/speaker
  - daemonset 형태로 배포되며, 외부 트래픽과 서비스를 연결해 네트워크 통신이 가능하도록 구성하는 역할을 담당합니다.
```

프로젝트에 적용된 톤앤 매너를 위해서 위와 같이 수정하면 좋을 것 같습니다.

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+MetalLB 설정은 ConfigMaps에 배포할 config 를 수정하여 설정 할 수 있습니다.
+
+MetalLB 에서 구성할 수 있는 모드로는 Layer 2 와 BGP 모드가 있습니다.
```
```suggestion
MetalLB 에서 구성할 수 있는 모드로는 다음과 같이 2가지가 있습니다.

1. [Layer 2 모드](https://metallb.universe.tf/concepts/layer2/)
2. [BGP 모드](https://metallb.universe.tf/concepts/bgp/)
```

추가적인 정보를 위해서 링크를 추가했습니다.

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+Layer 2 모드에서는 워커 노드의 네트워크 인터페이스에 IP 를 바인딩할 필요가 없습니다.
+
+로컬 네트워크의 ARP 요청에 직접 응답하여 컴퓨터의 MAC주소를 클라이언트에 제공하는 방식으로 작동합니다.
```
```suggestion
Layer 2 모드는 간단하게 사용할 IP 주소의 대역만 설정하면 됩니다.  
Layer 2 모드를 사용할 경우 워커 노드의 네트워크 인터페이스에 IP를 바인딩 하지 않아도 되는데 로컬 네트워크의 ARP 요청에 직접 응답하여 컴퓨터의 MAC주소를 클라이언트에 제공하는 방식으로 작동하기 때문입니다.
```

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+    - name: default
+      protocol: layer2
+      addresses:
+      - 192.168.35.100-192.168.35.110
```
```suggestion
      - 192.168.35.100-192.168.35.110  # IP 대역폭
```

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+### Kubeflow GUI
+
+먼저 kubeflow gui의 External-IP type 을 LoadBalancer 로 바꾸기전에 svc 를 확인합니다.
```
```suggestion
먼저 kubeflow gui의 External-IP type 을 LoadBalancer 로 바꾸기 전에 svc 를 확인합니다.
```

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
````diff
@@ -0,0 +1,280 @@
+NAME                   TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)                                        AGE
+istio-ingressgateway   ClusterIP   10.103.72.5   <none>        15021/TCP,80/TCP,443/TCP,31400/TCP,15443/TCP   4h21m
+
+```text
````
이 부분 위에 랩핑으로 수정 부탁드립니다.
suggestion이 안되네요 ㅠ

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+    istio: ingressgateway
+  sessionAffinity: None
+  type: LoadBalancer
+  loadBalancerIP: 192.168.35.100
```
```suggestion
  loadBalancerIP: 192.168.35.100  # Add IP
```

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+
+원하는 IP 주소를 입력하고 싶은 경우 loadBalancerIP 항목을 추가합니다.
+
+추가 안 할 경우에는 위에서 설정했던 IP 주소풀에서 순차적으로 IP 주소가 배정됩니다.
```
```suggestion
type 을 LoadBalancer 로 변경하고 원하는 IP 주소를 입력하고 싶은 경우 loadBalancerIP 항목을 추가합니다.  
추가 하지 않을 경우에는 위에서 설정한 IP 주소풀에서 순차적으로 IP 주소가 배정됩니다.
```

Aiden-Jeon commented on content/en/docs/appendix/metallb.md:
```diff
@@ -0,0 +1,280 @@
+    application-crd-id: kubeflow-pipelines
+  sessionAffinity: None
+  type: LoadBalancer
+  loadBalancerIP: 192.168.35.101
```
```suggestion
  loadBalancerIP: 192.168.35.101  # Add IP
```

Aiden-Jeon reviewed (commented):
프로젝트의 다른 글들과의 톤앤 매너를 위해 몇가지 코멘트 했습니다.
제가 영문 버전에 코멘트를 달았는데 국문 버전도 동일하게 부탁드립니다!
감사합니다!

kosehy commented:
> 프로젝트의 다른 글들과의 톤앤 매너를 위해 몇가지 코멘트 했습니다. 제가 영문 버전에 코멘트를 달았는데 국문 버전도 동일하게 부탁드립니다! 감사합니다!

@Aiden-Jeon 자세히 리뷰해 주셔서 너무 감사합니다! @Aiden-Jeon 님의 의견을 반영해서 다시 커밋하겠습니다.

kosehy committed:
docs: translate metallb.md to english

Translate metallb.md to english.
Fix one sentence in korean metallb.md for improving the readability

kosehy committed:
docs: fix no space issue

kosehy committed:
docs: correct grammer and typo using grammarly

Aiden-Jeon reviewed (approved):
LGTM!
감사합니다!
